### ISSUE TITLE
eval compiles to module.runModuleSetters(eval(...)) even in packages that do not use modules

### ISSUE THREAD

**Issue by MatejJan** [meteor/meteor#8365]
When placing `eval(...)` in a source file within a package that does not do `api.use('modules');` execution will break on the client with:

```
ReferenceError: module is not defined
```

Looking at compiled javascript, the eval line has become: `module.runModuleSetters(eval(...))`.

Since the package does not use modules I expect this is a bug.

The way to go around this issue is to include `api.use('modules');` in your `package.js`.

## Reproduction

[Reproduction repository](https://github.com/MatejJan/meteor-modules-eval-bug)

**`client/main.html`**
```
<body>
  Random number using eval: {{eval}}
</body>
```

**`client/main.js`**
```
import { Template } from 'meteor/templating';

Template.body.helpers({
  eval() {
    return Test.eval();
  }
});
```

**`packages/test/main.coffee`**
```
class Test
  @eval: ->
    eval 'Math.random()'
```

**`packages/test/package.js`**
```
Package.onUse(function(api) {
  api.use('coffeescript');

  // Uncomment this to make the project work.
  // api.use('modules');

  api.export('Test');
  api.addFiles('test.coffee');
});
```

I'm running the latest meteor (1.4.2.6) on macOS Sierra (10.12.3).

**Comment by benjamn** [comment]
I believe this is because the `coffeescript` plugin uses `babel-compiler` to compile `import` declarations, and `eval` is affected because it just might be updating the `exports` object.

There's been some [talk](https://github.com/meteor/meteor/pull/7033/commits/c4c8f826c3c4a66d4880b26d8f67868a53c1c52c) in the past of making `coffeescript/package.js` call `api.imply("modules")` so that anyone using `coffeescript` will also automatically have a CommonJS module system.

The problem you've identified here suggests it isn't fully true that you can use `coffeescript` without also using `modules`, so perhaps we should go ahead and tie them together?

What say you, @GeoffreyBooth @abernix @klaussner (& anyone else who has feelings)?

**Comment by GeoffreyBooth** [comment]
Remind me why we didn't make the plunge and imply modules before? If the only reason was because we didn't want to always pass the CoffeeScript compiler's output through Babel, that's not a good reason since 1.12 when we switched to always using Babel. Was there another reason? Just to avoid the overhead for projects that use CommonJS or auto-included files?

**Comment by benjamn** [comment]
I can guarantee you someone is going to report an issue where code behaves differently because `require`, `exports`, and/or `module` are defined. This issue (#8365) makes me lean towards accepting that risk (and just helping folks fix it).

**Comment by abernix** [comment]
I'm thinking 🤞 any uses of `module` and `exports` will be _mostly_-outside the bulk of their logic and less of an edge case.  So hopefully, things will fail ceremoniously enough that it might just be "ok" to let that failure happen.  Obviously I don't like the idea of breaking anything, but I think having `modules` would most-often be welcomed.

...then we can guide them to fix it.

**Comment by GeoffreyBooth** [comment]
This isn't *that* much of an edge case. Think of all the libraries that try to detect `require`, or require.js. Most such libraries will hopefully be installed via packages or NPM modules or the `client/compatibility` folder, so hopefully won't be an issue, but I'm sure it'll come up. Then again, you've probably already been dealing with this with the`ecmascript` package.

**Comment by mitar** [comment]
I think we should definitely not imply modules automatically. Why not do it in the opposite direction: if modules are not added to the package, then CoffeeScript does not go and use `babel-compiler` to process `import` declarations?

**Comment by benjamn** [comment]
@mitar Since you use `coffeescript` a lot, can you point us to an example of something that would break if `coffeescript` did `api.imply("modules")`?

Unfortunately, the `coffeescript` plugin has no way of knowing (at build time) whether an arbitrary package or application using `coffeescript` also uses `modules`.

**Comment by mitar** [comment]
Hm, and currently it is not possible to chain build tools? So that CoffeeScript would compile to normal JavaScript, and then babel would take it from there and transform that, processing imports? So if one would have both `coffeescript` and `ecmascript` as a dependency, `ecmascript` would post-process any JS file made by CoffeeScript. And if there is no `ecmascript`, the latter would not happen?

**Comment by mitar** [comment]
And I do not know what would break, I do not use `modules`, so I do not have experience of interaction between CoffeeScript and modules.

**Comment by GeoffreyBooth** [comment]
CoffeeScript needs Babel for more than just modules. Since generators were added a few versions ago, there are several ES2015 features that are opt-in and output as ES2015+.

**Comment by benjamn** [comment]
Currently `coffeescript` (the plugin) depends directly on `babel-compiler`, not `ecmascript`. That's how we "chain" compiler plugins today. We've talked about this before, but I'll reiterate that coordinating compiler plugin chains is a complicated problem, perhaps best exemplified by the Babel plugin configuration system.

Since you don't have concrete arguments against implying `modules`, this discussion is back to where it was before you chimed in.

**Comment by mitar** [comment]
True, this is why I think we should chain them together, no? If you opt-in, then you enable `ecmascript` as well?

One more idea. What if we make this an option for `addFiles`? `api.addFiles(['foo.coffee'], {ecmascript: true})` and in this case the output goes through Babel?

**Comment by mitar** [comment]
Also, why `api.imply` and not simply `api.use`? If I understand correctly, with `api.imply` in `coffeescript` package, if `coffeescript` is `api.use`d in another package X, then `modules` gets also activated for whole package X?

**Comment by benjamn** [comment]
> What if we make this an option for `addFiles`? `api.addFiles(['foo.coffee'], {ecmascript: true})` and in this case the output goes through Babel?

That doesn't work for `.coffee` files in apps, though, right?

> Also, why `api.imply` and not simply `api.use`? If I understand correctly, with `api.imply` in `coffeescript` package, if `coffeescript` is `api.use`d in another package X, then `modules` gets also activated for whole package X?

That's the intended difference between `api.use` and `api.imply`: package X doesn't depend on packages that its dependencies `api.use`, but it does depend on those they `api.imply`.

That said, there are definitely some packages that modify the global environment in a way that's visible to any package, so this distinction is sometimes lost.

**Comment by mitar** [comment]
> That won't work for .coffee files in apps, though, right?

I mean, we need a way to set options for files in apps anyway. We are waiting for the same thing for autoprefixer, which currently also works just for `addFiles`.

**Comment by hwillson** [comment]
Since this issue has some discussion about the implications of having `modules` implied when using `coffeescript`, I'll just drop a pointer to issue #8115 here (since it's related - whatever solution we decide on here will take care of both issues).

**Comment by mitar** [comment]
I use CoffeeScript a lot, but I checked many of my packages, and I must admit that when I have a package which wraps some NPM module (or another library), I in general just use JavaScript, and not CoffeeScript. So I am not sure how many packages are really out there which use CoffeeScript and are wrapping a NPM at the same time. And even if it exists, it can probably be split into two packages.

So I am not really sure how much would implying `modules` by default really break things. I do not like it, but that is for other reasons (loading extra code which I do not use, issues with packages being traversed as a whole #6823 - but if we would have #7598, this would satisfy me and I would have less against `modules` being everywhere).

**Comment by hwillson** [comment]
This is no longer an issue with Meteor 1.6, which uses CoffeeScript 2 (I've updated the supplied repro to 1.6 and verified). Thanks!

**Comment by GeoffreyBooth** [comment]
Was it fixed because of Meteor 1.6, or because of CoffeeScript 2 (or only with both together)? You can use Meteor 1.6 with CoffeeScript 1.

**Comment by hwillson** [comment]
@GeoffreyBooth It's fixed by Meteor 1.6, not CoffeeScript 2 (I forgot your CoffeeScript 2 changes aren't slated to be released until 1.6.1). So updating the repro to use Meteor 1.6 (which is still using `coffeescript@1.12.7_3` and `coffeescript-compiler@1.12.7_3`) is all that is needed.
